### ISSUE TITLE
Community Marketplace open-ended version range

### ISSUE THREAD

**Issue by Nadahar** [openhab/openhab-core#4523]
I find the current practice of always having to specify a compatibility end a bit strange. Since there's often no good reason to expect that something will stop working at a future version, the current "convention" seems to be to specify the next major version.

This means that now that `main` has changed to `5.0.0-SNAPSHOT` pretty much all add-ons with a version range are now hidden - and in reality, the vast majority are likely to work just fine with `5.0.0`. This means that a lot of Community Marketplace entries will have to be edited, or they will disappear from future versions. I think the upper bound only makes sense when there's a known reason why it would stop working.

Thus, I'm proposing to change the logic to accept an open-ended range in the form `[x.x.x(.x)?;)` or`[x.x.x(.x)?;]`.

The change necessary is small, but as I can't submit a PR, I'll just have to make a suggestion here:
```diff
 .../java/org/openhab/core/addon/marketplace/BundleVersion.java     | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/BundleVersion.java b/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/BundleVersion.java
index ee01b7b51..c9ac47a28 100644
--- a/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/BundleVersion.java
+++ b/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/BundleVersion.java
@@ -29,7 +29,7 @@ public class BundleVersion {
     private static final Pattern VERSION_PATTERN = Pattern.compile(
             "(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<micro>\\d+)(\\.((?<rc>RC)|(?<milestone>M))?(?<qualifier>\\d+))?");
     public static final Pattern RANGE_PATTERN = Pattern.compile(
-            "\\[(?<start>\\d+\\.\\d+(?<startmicro>\\.\\d+(\\.\\w+)?)?);(?<end>\\d+\\.\\d+(?<endmicro>\\.\\d+(\\.\\w+)?)?)(?<endtype>[)\\]])");
+            "\\[(?<start>\\d+\\.\\d+(?<startmicro>\\.\\d+(\\.\\w+)?)?);(?<end>\\d+\\.\\d+(?<endmicro>\\.\\d+(\\.\\w+)?)?)?(?<endtype>[)\\]])");
 
     private final String version;
     private final int major;
@@ -87,7 +87,10 @@ public class BundleVersion {
             return false;
         }
 
-        String endString = matcher.group("endmicro") != null ? matcher.group("end") : matcher.group("stop") + ".0";
+        if (matcher.group("end") == null) {
+            return true;
+        }
+        String endString = matcher.group("endmicro") != null ? matcher.group("end") : matcher.group("end") + ".0";
         boolean inclusive = "]".equals(matcher.group("endtype"));
         BundleVersion endVersion = new BundleVersion(endString);
         int comparison = this.compareTo(endVersion);
```
I've tested the above code, and it seems to work just fine.

Please note that there's also a bug fix here, since `matcher.group("stop")` throws an exception it that code is ever reached, since there's no `stop` group defined `RANGE_PATTERN`. I'm pretty sure it should be `end` instead of `stop`.

**Comment by lolodomo** [comment]
I think the idea is to have marketplace versions that were compiled for the target server.
We all have in mind users complaining in the community forum about an unstable OH server until we discovered it was due to a very old binding version installed from the marketplace. This is even the first thing to check when the OH server is not working well.
Each OH version is coming with changes in the core framework.

Oh 5.0 will come with a major change: the move to 64 bit.

**Comment by Nadahar** [comment]
Yes, this isn't meant for bindings, but for other type of add-ons where there's practically no way to predict when it might "become incompatible". Things like UI widgets, rule templates etc. In other words, add-ons that aren't Java code.

I see this as a big problem, because the authors will just have to "guess" what the upper bound should be. Many forget about this, and a lot of the current add-ons on the marketplace that works just fine are thus excluded. Another thing is that it's still very much possible to create an add-on with no version range at all, working with all versions of OH. So, when you can't have an open upper bound, many simply choose to get around this by dropping the version range - which will make it available to all versions.

In addition, as mentioned, there is a bug fix here. The current code throws an exception because it's using a non-existing regex group.

But, since this issue was created a long time ago now, a lot has happened since then, and I've reimplemented this whole logic since then, in code that for other reasons probably won't be submitted. Regardless, I'm closing this as the other implementation would sole many more problems.
